### ISSUE TITLE
QGIS 3.28 crash when loading this gpkg

### ISSUE THREAD

**Issue by rduivenvoorde** [qgis/QGIS#55488]
### What is the bug or the crash?

When I load the layer 'plaatsen' from this gpkg: https://duif.net/nl.gpkg a recent QGIS 3.28 crashes.

Tested both on Debian Bookworm (self compiled) as on Debian testing (idem).
2 different machines, clean profiles

Loading other gpkg's are fine.

So it could very much be this specific old demo gpkg (created with a probably ancient QGIS version...)

I hope others can reproduce this.

### Steps to reproduce the issue

- download https://duif.net/nl.gpkg and safe to /tmp
- in QGIS browser(!), browse to /tmp, open the nl.gpkg node and load plaatsen by double-clicking on it

QGIS crashes with this info:

```
[New LWP 3909]
[New LWP 3935]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
0x00007f105f2f11b7 in __GI___wait4 (pid=4018, stat_loc=0x7ffeb5d49704, options=0, usage=0x0) at ../sysdeps/unix/sysv/linux/wait4.c:30
30	../sysdeps/unix/sysv/linux/wait4.c: No such file or directory.
[Current thread is 1 (Thread 0x7f1049cd6840 (LWP 3879))]
#0  0x00007f105f2f11b7 in __GI___wait4 (pid=4018, stat_loc=0x7ffeb5d49704, options=0, usage=0x0) at ../sysdeps/unix/sysv/linux/wait4.c:30
        sc_ret = -512
        sc_cancel_oldtype = 0
#1  0x000055fe7b786670 in qgisCrash(int) (signal=-1) at /home/richard/git/qgis/src/app/main.cpp:349
        status = 32766
        pidstr = "--pid=3879\000{\376U\000\000X\226Ե\376\177\000\000\300\024'_\020\177\000"
        gdbpid = 4018
        exename = "/home/richard/git/qgis/build/output/bin/qgis\000\177", '\000' <repeats 18 times>, "p\225Ե\376\177\000\000p\225Ե\376\177\000\000\360\225Ե\376\177\000\000{\000\000\000\000\000\000\000\004\000\000\000\020\177\000\000\240&?_\020\177\000\000Fatal: ASSERT: \"ds->refCount > 0\" in file /home/richard/git/qgis/src/core/providers/ogr/qgs"...
        len = 44
#2  0x000055fe7b786cf7 in myMessageOutput(QtMsgType, QMessageLogContext const&, QString const&) (type=QtFatalMsg, msg=...) at /home/richard/git/qgis/src/app/main.cpp:446
#3  0x00007f105bed26b0 in  () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#4  0x00007f105bed27c8 in  () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#5  0x00007f105be910ba in QMessageLogger::fatal(char const*, ...) const () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#6  0x00007f105be902ca in qt_assert_x(char const*, char const*, char const*, int) () at /lib/x86_64-linux-gnu/libQt5Core.so.5
#7  0x00007f1059d444be in QgsOgrProviderUtils::getLayer(QString const&, QString const&, QString&) (dsName=..., layerName=..., errCause=...) at /home/richard/git/qgis/src/core/providers/ogr/qgsogrproviderutils.cpp:1627
        iter2 = {i = 0x55fe82c5d2e0}
        ds = 0x55fe80efd4e0
        __for_range = @0x7ffeb5d49a60: {<QListSpecialMethods<QgsOgrProviderUtils::DatasetWithLayers*>> = {<No data fields>}, {p = {static shared_null = <error reading variable: Cannot access memory at address 0x7f105c1c8120>, d = 0x55fe81289750}, d = 0x55fe81289750}}
        __for_begin = {i = 0x55fe81289760}
        __for_end = {i = 0x55fe81289768}
        datasetList = <error reading variable: Cannot access memory at address 0x55fe82c32550>
        constDatasetList = {<QListSpecialMethods<QgsOgrProviderUtils::DatasetWithLayers*>> = {<No data fields>}, {p = {static shared_null = <error reading variable: Cannot access memory at address 0x7f105c1c8120>, d = 0x55fe81289750}, d = 0x55fe81289750}}
        iter = {i = 0x55fe82c32520}
        locker = {val = 139708235570969}
#8  0x00007f1059d21b6a in LoadDataSourceLayerStylesAndLayer(QString const&, QgsOgrLayerUniquePtr&, QgsOgrLayerUniquePtr&, QString&) (uri=Python Exception <class 'gdb.MemoryError'>: Cannot access memory at address 0x7ffeb5d49b08
#9  0x00007f1059d21d2a in QgsOgrProviderMetadata::loadStyle(QString const&, QString&) (this=0x55fe7d067cb0, uri=..., errCause=...) at /home/richard/git/qgis/src/core/providers/ogr/qgsogrprovidermetadata.cpp:679
        layerStyles = Python Exception <class 'gdb.MemoryError'>: Cannot access memory at address 0x7ffeb5d49c18
#10 0x00007f1059c748f5 in QgsProviderRegistry::loadStyle(QString const&, QString const&, QString&) (this=Python Exception <class 'gdb.MemoryError'>: Cannot access memory at address 0x7ffeb5d49d10
#11 0x00007f105a8221f8 in QgsVectorLayer::loadNamedStyle(QString const&, bool&, bool, QFlags<QgsMapLayer::StyleCategory>) (this=Python Exception <class 'gdb.MemoryError'>: Cannot access memory at address 0x7ffeb5d49d90
#12 0x00007f105a821c58 in QgsVectorLayer::loadNamedStyle(QString const&, bool&, QFlags<QgsMapLayer::StyleCategory>) (this=0x55fe80eed4d0, theURI=..., resultFlag=<error reading variable: Cannot access memory at address 0x7ffeb5d49fdf>, categories=...) at /home/richard/git/qgis/src/core/vector/qgsvectorlayer.cpp:5504
#13 0x00007f1059f8100f in QgsMapLayer::loadDefaultStyle(bool&) (this=0x55fe80eed4d0, resultFlag=<error reading variable: Cannot access memory at address 0x7ffeb5d49fdf>) at /home/richard/git/qgis/src/core/qgsmaplayer.cpp:1052
#14 0x00007f105a7fcb56 in QgsVectorLayer::loadDefaultStyle(bool&) (this=Python Exception <class 'gdb.MemoryError'>: Cannot access memory at address 0x7ffeb5d49f50
#15 0x00007f105fd1dbf3 in QgsAppLayerHandling::postProcessAddedLayer(QgsMapLayer*) (layer=0x55fe80eed4d0) at /home/richard/git/qgis/src/app/layers/qgsapplayerhandling.cpp:75
        ok = Python Exception <class 'gdb.MemoryError'>: Cannot access memory at address 0x7ffeb5d49fdf
#16 0x00007f105fd23170 in QgsAppLayerHandling::addSublayers(QList<QgsProviderSublayerDetails> const&, QString const&, QString const&) (layers=..., baseName=..., groupName=...) at /home/richard/git/qgis/src/app/layers/qgsapplayerhandling.cpp:766
        ml = Python Exception <class 'gdb.MemoryError'>: Cannot access memory at address 0x7ffeb5d4a2d0
#17 0x00007f105fd2f378 in QgsAppLayerHandling::addLayerPrivate<QgsVectorLayer>(QgsMapLayerType, QString const&, QString const&, QString const&, bool) (type=Python Exception <class 'gdb.MemoryError'>: Cannot access memory at address 0x7ffeb5d4a36c
#18 0x00007f105fd25aeb in QgsAppLayerHandling::addVectorLayer(QString const&, QString const&, QString const&) (uri=Python Exception <class 'gdb.MemoryError'>: Cannot access memory at address 0x7ffeb5d4a7d8
Killed
```

### Versions

QGIS version | 3.28.13-Firenze | QGIS code revision | 58aa7b8549c
-- | -- | -- | --
Qt version | 5.15.10
Python version | 3.11.6
GDAL/OGR version | 3.8.0
PROJ version | 9.3.0
EPSG Registry database version | v10.094 (2023-08-08)
GEOS version | 3.12.1-CAPI-1.18.1
SQLite version | 3.44.2
PostgreSQL client version | unknown
SpatiaLite version | 5.1.0
QWT version | 6.1.4
QScintilla2 version | 2.14.1
OS version | Debian GNU/Linux trixie/sid

### Supported QGIS version

- [X] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [X] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

**Comment by agiudiceandrea** [comment]
The issue doesn't occur using QGIS 3.28.13 (OSGeo4W) on Windows 10.

**Comment by rduivenvoorde** [comment]
@agiudiceandrea did you try to compile on Windows? Because it is only on latest 3.28.13, not on that release (my fear is that 3.28.14 will crash, not 3.28.13)

**Comment by agiudiceandrea** [comment]
I cannot replicate the issue on my Windows 10 system neither using QGIS 3.28.13 (3b4db4f561) from OSGeo4W (qgis-ltr)

```
QGIS version
3.28.13-Firenze
QGIS code revision
3b4db4f561
Qt version
5.15.3
Python version
3.9.5
GDAL/OGR version
3.8.1
PROJ version
9.3.1
EPSG Registry database version
v10.098 (2023-11-24)
GEOS version
3.12.1-CAPI-1.18.1
SQLite version
3.41.1
PDAL version
2.6.0
PostgreSQL client version
unknown
SpatiaLite version
5.1.0
QWT version
6.1.6
QScintilla2 version
2.13.4
OS version
Windows 10 Version 1903

```
nor using QGIS 3.28.13 (58aa7b8549c) from OSGeo4W (qgis-ltr-dev)


```

QGIS version
3.28.13-Firenze
QGIS code revision
58aa7b85
Qt version
5.15.3
Python version
3.9.5
Compiled against GDAL/OGR
3.8.1
Running against GDAL/OGR
3.8.0
PROJ version
9.3.1
EPSG Registry database version
v10.098 (2023-11-24)
GEOS version
3.12.1-CAPI-1.18.1
SQLite version
3.41.1
PDAL version
2.6.0
PostgreSQL client version
unknown
SpatiaLite version
5.1.0
QWT version
6.1.6
QScintilla2 version
2.13.4
OS version
Windows 10 Version 1903

This copy of QGIS writes debugging output.

```

**Comment by rouault** [comment]
I believe the crash is specifc to running a QGIS debug build with assertions turned on, and it might have been fixed per 9963cb5095f8

**Comment by rduivenvoorde** [comment]
@rouault thanks! I think you are right. Compiled a fresh master and now all is OK!
